### ISSUE TITLE
Coremod injections into a class (ClientPacketListener) break mixins for the same class

### ISSUE THREAD

**Issue by thexaero** [MinecraftForge/MinecraftForge#8356]
**Minecraft Version:** 
1.18.1

**Forge Version:** 
39.0.10

**Logs:** 
Compatibility issue between Xaero's Minimap and Origins Enhanced:
```
[main/WARN] [mixin/]: Error loading class: net/minecraft/client/multiplayer/ClientPacketListener (java.lang.UnsupportedOperationException: PermittedSubclasses requires ASM9)
[main/WARN] [mixin/]: @Mixin target net.minecraft.client.multiplayer.ClientPacketListener was not found originsenhanced-common.mixins.json:client.multiplayer.ClientPacketListenerMixin
```

**Steps to Reproduce:**
 1. Install [Xaero's Minimap](https://www.curseforge.com/minecraft/mc-mods/xaeros-minimap/files/3582980)
 2. Install [Origins Enhanced](https://www.curseforge.com/minecraft/mc-mods/origins-enhanced/files/3571710)
 3. Wait for the game to load (title screen is enough) and check the console/logs.

**Description of issue:** 
It doesn't seem to matter what I inject into a class (in my case `ClientPacketListener`) with a coremod, it always breaks the mixin for the same class from another mod (in my case Origins Enhanced). Even transformers that don't do anything cause the issue. I have not tested other classes or mods. My best guess is that the issue is on Forge's end.

**Comment by TheCurle** [comment]
I'll prepare a test case.  
It may be worth reporting this to the Mixin bug tracker too.

**Comment by Mumfrey** [comment]
Nope, still McModLauncher/modlauncher#74

**Comment by TheCurle** [comment]
Damnit. I'll note down this error for future reference.

We really need to get McModLauncher operational soon..

**Comment by thexaero** [comment]
@Mumfrey Tyvm for your input! Way too confusing for 2 am, ngl. Does this mean that the Origins mod is referencing a sealed class in their ClientPacketListener mixin? Does the coremod transformer cause an older api version (ASM7?) to be passed to ASM?

**Comment by Mumfrey** [comment]
It's not the coremod transformer per se, it's that API constant in modlauncher. It cannot interact with records or sealed classes. 

Mixin keeps getting clobbered with this because often it's the first thing which tries to drag something through the transformer chain (during generation of metadata) so the CNFE gets caught by `ClassInfo` and then propagated as a `null`, so the log message you see from mixin is from [ClassInfo->forName()](https://github.com/SpongePowered/Mixin/blob/master/src/main/java/org/spongepowered/asm/mixin/transformer/ClassInfo.java#L2009) when it catches the upstream error. Note from the code that the full exception is available at TRACE.

**Comment by thexaero** [comment]
Thanks for the useful info! 
Now that I have looked at the full stack trace and went through a lot of the Mixin and McModLoader code, I think I have a somewhat better understanding of the issue.
```
java.lang.UnsupportedOperationException: PermittedSubclasses requires ASM9
	at org.objectweb.asm.ClassVisitor.visitPermittedSubclass(ClassVisitor.java:266) ~[asm-9.1.jar:?]
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:683) ~[asm-9.1.jar:?]
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:401) ~[asm-9.1.jar:?]
	at cpw.mods.modlauncher.TransformerClassWriter.computeHierarchyFromFile(TransformerClassWriter.java:151) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.modlauncher.TransformerClassWriter.computeHierarchy(TransformerClassWriter.java:112) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.modlauncher.TransformerClassWriter$SuperCollectingVisitor.lambda$visit$0(TransformerClassWriter.java:185) ~[modlauncher-9.0.7.jar:?]
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992) ~[?:?]
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762) ~[?:?]
	at cpw.mods.modlauncher.TransformerClassWriter$SuperCollectingVisitor.visit(TransformerClassWriter.java:184) ~[modlauncher-9.0.7.jar:?]
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:546) ~[asm-9.1.jar:?]
	at org.objectweb.asm.ClassReader.accept(ClassReader.java:401) ~[asm-9.1.jar:?]
	at cpw.mods.modlauncher.TransformerClassWriter.computeHierarchyFromFile(TransformerClassWriter.java:151) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.modlauncher.TransformerClassWriter.computeHierarchy(TransformerClassWriter.java:112) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.modlauncher.TransformerClassWriter.getSupers(TransformerClassWriter.java:83) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.modlauncher.TransformerClassWriter.getCommonSuperClass(TransformerClassWriter.java:63) ~[modlauncher-9.0.7.jar:?]
	at org.objectweb.asm.SymbolTable.addMergedType(SymbolTable.java:1202) ~[asm-9.1.jar:?]
	at org.objectweb.asm.Frame.merge(Frame.java:1299) ~[asm-9.1.jar:?]
	at org.objectweb.asm.Frame.merge(Frame.java:1197) ~[asm-9.1.jar:?]
	at org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1610) ~[asm-9.1.jar:?]
	at org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1546) ~[asm-9.1.jar:?]
	at org.objectweb.asm.tree.MethodNode.accept(MethodNode.java:769) ~[asm-tree-9.1.jar:?]
	at org.objectweb.asm.tree.MethodNode.accept(MethodNode.java:649) ~[asm-tree-9.1.jar:?]
	at org.objectweb.asm.tree.ClassNode.accept(ClassNode.java:452) ~[asm-tree-9.1.jar:?]
	at cpw.mods.modlauncher.ClassTransformer.transform(ClassTransformer.java:133) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.modlauncher.TransformingClassLoader.maybeTransformClassBytes(TransformingClassLoader.java:50) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.cl.ModuleClassLoader.getMaybeTransformedClassBytes(ModuleClassLoader.java:242) ~[securejarhandler-0.9.54.jar:?]
	at cpw.mods.modlauncher.TransformingClassLoader.buildTransformedClassNodeFor(TransformingClassLoader.java:58) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.modlauncher.LaunchPluginHandler.lambda$announceLaunch$10(LaunchPluginHandler.java:100) ~[modlauncher-9.0.7.jar:?]
	at org.spongepowered.asm.launch.MixinLaunchPluginLegacy.getClassNode(MixinLaunchPluginLegacy.java:222) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.launch.MixinLaunchPluginLegacy.getClassNode(MixinLaunchPluginLegacy.java:207) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.ClassInfo.forName(ClassInfo.java:2005) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinInfo.getTargetClass(MixinInfo.java:1017) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinInfo.readTargetClasses(MixinInfo.java:1007) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinInfo.parseTargets(MixinInfo.java:895) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinConfig.prepareMixins(MixinConfig.java:867) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinConfig.prepare(MixinConfig.java:779) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.prepareConfigs(MixinProcessor.java:539) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.select(MixinProcessor.java:462) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.checkSelect(MixinProcessor.java:438) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinProcessor.applyMixins(MixinProcessor.java:290) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.mixin.transformer.MixinTransformer.transformClass(MixinTransformer.java:250) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.service.modlauncher.MixinTransformationHandler.processClass(MixinTransformationHandler.java:131) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at org.spongepowered.asm.launch.MixinLaunchPluginLegacy.processClass(MixinLaunchPluginLegacy.java:131) ~[mixin-0.8.5.jar:0.8.5+Jenkins-b310.git-155314e6e91465dad727e621a569906a410cd6f4]
	at cpw.mods.modlauncher.serviceapi.ILaunchPluginService.processClassWithFlags(ILaunchPluginService.java:156) ~[modlauncher-9.0.7.jar:9.0.7+91+master.8569cdf]
	at cpw.mods.modlauncher.LaunchPluginHandler.offerClassNodeToPlugins(LaunchPluginHandler.java:88) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.modlauncher.ClassTransformer.transform(ClassTransformer.java:120) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.modlauncher.TransformingClassLoader.maybeTransformClassBytes(TransformingClassLoader.java:50) ~[modlauncher-9.0.7.jar:?]
	at cpw.mods.cl.ModuleClassLoader.readerToClass(ModuleClassLoader.java:110) ~[securejarhandler-0.9.54.jar:?]
	at cpw.mods.cl.ModuleClassLoader.lambda$findClass$16(ModuleClassLoader.java:213) ~[securejarhandler-0.9.54.jar:?]
	at cpw.mods.cl.ModuleClassLoader.loadFromModule(ModuleClassLoader.java:223) ~[securejarhandler-0.9.54.jar:?]
	at cpw.mods.cl.ModuleClassLoader.findClass(ModuleClassLoader.java:213) ~[securejarhandler-0.9.54.jar:?]
```

 Here is how I see it:
1) Mixin starts side-loading Mixin target classes, one of which is `ClientPacketListener`
2) One of [these](https://github.com/McModLauncher/modlauncher/blob/6bb4b073b13cc89477aa2585756143c97d6a29a1/src/main/java/cpw/mods/modlauncher/ClassTransformer.java#L122) saves the day, when there are no coremod transformers like mine. Otherwise, [this](https://github.com/McModLauncher/modlauncher/blob/6bb4b073b13cc89477aa2585756143c97d6a29a1/src/main/java/cpw/mods/modlauncher/ClassTransformer.java#L133) has the potential to break everything as it has ASM7 passed to it, throwing the provided exception while recomputing frames for `ClientPacketListener` methods. 

What I don't understand is why doesn't my mod cause the same issue without Origins being in play, or why does the Origins mod work fine without my mod in play. Either of them alone should cause frames to recompute for `ClientPacketListener` methods simply because they are both transforming it. Why does it only affect Mixin side-loading `ClientPacketListener`? Am I missing something?

**Comment by sciwhiz12** [comment]
Fixed by 00fafb07cee8aea39e09a4e8189c2fefcbed9eca (published as 1.18.1-39.0.44) with the bump of ModLauncher version to v9.1.0, which includes McModLauncher/modlauncher#88.
